### PR TITLE
Infer GitHub user Twitter handle

### PR DESCRIPTION
**Pull request by bevacqua**
Here's a fix for #46. Uses [an external API service](https://github.com/bevacqua/github-npm-profile) because the dependencies don't work browserified _(and would be too large anyways, `cheerio` is among them for instance)_

- Only runs on user profiles
- If user already has Twitter links in profile they are beautified
  - This works regardless of them having a configured npm profile
  - No requests are sent to the service in this case
- Looks up GitHub user on npm
- Pulls Twitter handle from npm profile
- Removes any previous vcard details which included a link to their Twitter handle
- Appends Twitter handle link before the last item in the info list which is always the join date

Before:

screen shot 2016-03-04 at 13 08 05

After:

screen shot 2016-03-04 at 14 14 28

### CODE REVIEW

**Comment by hkdobrev** [comment]
:-1: I'm sorry, but personally I don't see the benefit of this. It would make a request on every GitHub profile, it would work only for users of npm and still even if it works in all cases, I think it is rare that you'll need it and you could do it manually. E.g. many people would either use the GitHub website field for their Twitter profile or they'll link to their Twitter from their website.

**Comment by hkdobrev** [comment]
I'd suggest it to be a separate extension and everyone who needs it could install it.

**Comment by SpaceK33z** [comment]
I agree with @hdkobrev, the combination of minimal benefit and that it would only work for npm users does it for me.

**Comment by bevacqua** [comment]
Updated.

- If user already has Twitter links in profile they get replaced and no requests are sent at all
  - This works regardless of them having a configured npm profile
- Fixed XSS vector

File: extension/content.js [hunk 1]
```diff
@@ -227,6 +228,32 @@ document.addEventListener('DOMContentLoaded', () => {
 		window.addEventListener('resize', infinitelyMore);
 	}
 
+	if (isProfile) {
```
Comment (hkdobrev) [inline extension/content.js]: I think `isProfile` would match both user profiles and organisation profiles. You might want to check if this is indeed a user profile page based on the DOM.

Comment (bevacqua) [inline extension/content.js]: Fixed

File: extension/content.js [hunk 2]
```diff
@@ -227,6 +227,33 @@ document.addEventListener('DOMContentLoaded', () => {
 		window.addEventListener('resize', infinitelyMore);
 	}
 
+	const details = $(`.vcard-details .vcard-detail`);
```
Comment (sindresorhus) [inline extension/content.js]: No point in using template literal here.

File: extension/content.js [hunk 3]
```diff
@@ -227,6 +227,33 @@ document.addEventListener('DOMContentLoaded', () => {
...
+		existing.forEach(x => {
+			const $el = $(x.el);
+			addTwitterProfile(x.matches[1], $el);
+			$el.remove();
+		});
+
+		if (!existing.length) {
+			fetch(`https://github-npm-profile.herokuapp.com/${ownerName}`)
+				.then(res => res.json())
+				.then(profile => sanitizedTwitterHandle(profile))
```
Comment (sindresorhus) [inline extension/content.js]: Sanitizing should be done server-side.

**Comment by sindresorhus** [comment]
I like this. I often need to look up people's Twitter account.

**Comment by sindresorhus** [comment]
> It would make a request on every GitHub profile

It's just a tiny request in the background though. You wouldn't know or care.

> it would work only for users of npm and still even if it works in all cases

That's true, but personally, most of the profiles I visit have a npm profile. And the server can easily be improved to support additional services.

File: extension/content.js [hunk 4]
```diff
@@ -227,6 +227,33 @@ document.addEventListener('DOMContentLoaded', () => {
...
+			.filter(x => x.matches);
+
+		existing.forEach(x => {
+			const $el = $(x.el);
+			addTwitterProfile(x.matches[1], $el);
+			$el.remove();
+		});
+
+		if (!existing.length) {
+			fetch(`https://github-npm-profile.herokuapp.com/${ownerName}`)
```
Comment (sindresorhus) [inline extension/content.js]: Maybe name it something more generic, like `get-twitter-username`.

**Comment by sindresorhus** [comment]
Many people have their email in their GitHub profile, is there any way to automatically match the email to a Twitter account?

**Comment by sindresorhus** [comment]
It currently fails in some cases though: https://github.com/paulirish/, which is weird, since Paul has the exact same npm username: https://www.npmjs.com/~paulirish

File: extension/content.js (hunk 3 above)
Comment (bevacqua) [inline extension/content.js]: Sure, but they were worried about the service I made becoming malicious and injecting stuff on the page, hence client-side sanitization

Comment (SpaceK33z) [inline extension/content.js]: The problem is that you shouldn't insert a value from a random server directly in html. By using `.text(handle)` in `addTwitterProfile` it is automatically properly escaped. The way you fixed it is also possible, but I think `.text()` is cleaner.

Comment (bevacqua) [inline extension/content.js]: It was just easier this way, I don't have a preference either way.

File: extension/content.js (hunk 4 above)
Comment (bevacqua) [inline extension/content.js]: I like that. I'll look into other ways of figuring out their Twitter handle

File: extension/content.js (hunk 2 above)
Comment (bevacqua) [inline extension/content.js]: https://twitter.com/nzgb/status/663701680958083072

**Comment by bevacqua** [comment]
A way to infer Twitter users from email would be ideal, as it'd be much easier to get a GitHub user's email address (usually right in their profile, easy to infer from commits, etc).

The underlying [`find-npm-by-github`](https://github.com/finnp/find-npm-by-github/blob/master/index.js) leaves a bit to desired (it's pretty slow too).

File: extension/content.js (hunk 3 above)
Comment (sindresorhus) [inline extension/content.js]: > but I think .text() is cleaner.

:+1:

**Comment by joshmanders** [comment]
While I like the idea of having a twitter on the profile, I don't see any clean way to do it without enforcing arbitrary hacks on the user themselves. Personally I think this would better suited as something that GitHub themselves implement.

**Comment by sindresorhus** [comment]
> Personally I think this would better suited as something that GitHub themselves implement.

Most things in this extension would be better implemented by GitHub, but I think it's clear by now that they won't, so it is what it is.

**Comment by joshmanders** [comment]
I've never seen anyone put their twitter anywhere but in their homepage, so other than detecting a twitter url and swapping the icon out for a twitter icon, I don't see a use for this unless you start making requests other places such as npm to match the github usernames to see if they have twitter there. As a user, I'd prefer not to have a bunch of extra network calls when I am browsing on github just to get a users twitter handle.

**Comment by bevacqua** [comment]
You're just being idealistic. How often do you go to a user's profile page? How often do you not want some extra help finding out their Twitter handle if you like their work? A single request in user profile pages isn't going to have any visible impact to you as a user, regardless of whether you feel it's dirty as an engineer.

**Comment by joshmanders** [comment]
Actually I spend half of my day on GitHub and going to users pages is a good portion of that. I'd say in an average day I view about 50 profiles. That's 50 extra requests over my network that could crap out and for what? A twitter handle? What if npm goes down, now I can't use Github because the plugin breaks on a network request.

Honestly, I never have issues doing a quick check on npm or even twitter for a users handle matching their github handle.. Not many people have different handles across all services.

**Comment by hkdobrev** [comment]
I second what @joshmanders said.

I visit a lot of GitHub user profiles for _hiring research_ if I may call it that. They either don't have Twitter (majority of them) or if they do, they have already put it either in the website field or on their website which I've already opened anyway. Sometimes I'd even find them on Twitter first and then open their GitHub profile so this would be useful to me on very rare occasions. On top of that many people I'm looking at are not publishing packages on npm.

These are just my thoughts and my workflow of course. I don't have a say in what should go into this extension or not. I'm just trying to show it may not be a common case and it may not be so useful to many people.

As I've said above I suggest this to be a separate extension. It could be installed by those who find it useful. The other things in Refined GitHub do not perform external network requests and just improve the UX with things GitHub already have one way or the other.

File: extension/content.js (hunk 4 above)
Comment (hkdobrev) [inline extension/content.js]: @bevacqua You may try the Twitter API and see if there's a Twitter user with the same handle as the user on GitHub.

**Comment by sindresorhus** [comment]
> That's 50 extra requests over my network

One tiny request (12 bytes) out of 48 requests (745 KB) done on a profile page.

> What if npm goes down, now I can't use Github because the plugin breaks on a network request.

It shouldn't. It's all done async and gracefully.

> On top of that many people I'm looking at are not publishing packages on npm.

You're probably right about that, but nothing is stopping the server part from supporting more services and ways to get the Twitter handle in the future.

**Comment by sindresorhus** [comment]
That being said, I think it would make more sense to do a separate extension for it, for the simple reason that more people would benefit from it, not just people using `refined-github`.

I would be happy to link to it in the readme and help discuss ways to improve the Twitter handle inference.

**Pull request closed without merging**